xapian-applications/omega/handler_libmwaw.cc

```diff
@@ -0,0 +1,264 @@
+	if (p)
+	    eol = p++;
+	else
+	    p = eol = end;
+	if ((end - start) > 5 && memcmp(start, "meta:", 5) == 0) {
+	    start += 5;
+	    switch (*start) {
+		case 'i': {
+		    if (author.empty())
+			PARSE_FIELD(start, eol, "initial-creator", author);
+		    break;
+		}
+		case 'k': {
+		    PARSE_FIELD(start, eol, "keyword", keywords);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 3 && memcmp(start, "dc:", 3) == 0) {
+	    start += 3;
+	    switch (*start) {
+		case 'c': {
+		    if (!author.empty())
+			author.clear();
+		    PARSE_FIELD(start, eol, "creator", author);
+		    break;
+		}
+		case 's': {
+		    PARSE_FIELD(start, eol, "subject", keywords);
+		    break;
+		}
+		case 't': {
+		    PARSE_FIELD(start, eol, "title", title);
+		    break;
+		}
+	    }
+	} else if ((end - start) > 8 && memcmp(start, "dcterms:", 8) == 0) {
+	    start += 8;
+	    PARSE_FIELD(start, eol, "available", keywords);
+	}
+    }
+}
+
+static bool
+extract_word(string& dump,
+	     string& title,
+	     string& keywords,
+	     string& author,
+	     string& error,
+	     RVNGFileStream* input)
+{
+    RVNGString content_dump;
+    RVNGTextTextGenerator content(content_dump, false);
+    if (MWAWDocument::parse(input, &content) == MWAWDocument::MWAW_R_OK) {
+	dump = content_dump.cstr();
+    } else {
+	error = "Libmwaw Error: Failed to extract text";
+	return false;
+    }
+
+    RVNGString metadata_dump;
+    RVNGTextTextGenerator metadata(metadata_dump, true);
+    if (MWAWDocument::parse(input, &metadata) == MWAWDocument::MWAW_R_OK) {
+	const char* data = metadata_dump.cstr();
+	size_t len = metadata_dump.size();
+	parse_metadata(data, len, author, title, keywords);
+    } else {
+	error = "Libmwaw Error: Failed to extract metadata";
+	return false;
+    }
+    return true;
+}
+
+static bool
+extract_spreadsheet(string& dump,
+		    string& title,
+		    string& keywords,
+		    string& author,
+		    string& error,
+		    string& pages,
+		    RVNGFileStream* input)
+{
+    RVNGStringVector pages_metadata;
+    RVNGTextSpreadsheetGenerator metadata(pages_metadata, true);
+    MWAWDocument::Result result = MWAWDocument::parse(input, &metadata);
+    if (result != MWAWDocument::MWAW_R_OK) {
+	error = "Libmwaw Error: Failed to extract metadata";
+	return false;
+    } else {
+	string meta_temp;
+	for (unsigned i = 0; i < pages_metadata.size(); ++i) {
+	    meta_temp.append(pages_metadata[i].cstr());
+	}
+	const char* data = meta_temp.c_str();
+	size_t len = meta_temp.size();
+	parse_metadata(data, len, author, title, keywords);
+    }
+
+    RVNGStringVector pages_content;
+    RVNGTextSpreadsheetGenerator content(pages_content, false);
+    result = MWAWDocument::parse(input, &content);
+    if (result != MWAWDocument::MWAW_R_OK) {
+	error = "Libmwaw Error: Failed to extract text";
+	return false;
+    } else {
+	for (unsigned i = 0; i < pages_content.size(); ++i) {
+	    dump.append(pages_content[i].cstr());
+	}
+	pages = pages_content.size();
```

Comment:
What type does `pages_content.size()` return?

Here you assign it to `pages`, which is a `std::string&` but above you compare `i < pages_content.size()` where `i` is unsigned.  I'm a bit surprised one doesn't fail to compile, and am concerned that at least one isn't doing what we want it to...

Reply:
As per my understanding, `pages_content.size()` returns an integer (numerical value). 
`pages` in the handlers is supposed to store the number of the pages in the file, right?

So what I did was - iterate over the pages (pages_content.size() times) of the file and append the data to `dump`.
And later on line 175, I store the number of pages (`pages_content.size()`) in the document in the variable `pages`.

Am I interpreting things wrongly?

Reply:
> `pages` in the handlers is supposed to store the number of the pages in the file, right?

Yes, but it's a `std::string&` and C++ doesn't implicitly convert an integer to a string for you.

It looks (from `/usr/include/librevenge-0.0/librevenge/RVNGStringVector.h`) like `size()` returns `unsigned` here.

Testing that:

```
$ cat string-from-unsigned.cc 
#include <iostream>
#include <string>

void f(std::string& pages, unsigned size) {
    pages = size;
}

int main() {
    std::string s;
    f(s, 42);
    std::cout << "'" << s << "'\n";
}
$ g++ -Wall -W -O2 -s string-from-unsigned.cc
$ ./a.out
'*'
```

So the code does compile, but my concern that it isn't doing what we want is well founded - it actually sets `pages` to a single character based on the value assigned!

You want `pages = str(pages_content.size());` instead.

Reply:
Thank you for pointing this out. I have made the necessary changes.